Correctly handle duplicate arguments to AOTAutograd

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85301

If we do not deduplicate them, our custom autograd function
will double-count the gradient computed for the variable (since
the same x.grad field will be embedded into the graph twice.)

The alternative is to destroy aliasing relationships the inputs
and trace accumulating individual gradients for each of the
inputs into separate grad fields, but this prevents resizing
of inputs inside the traced graph from working correctly.  In
principle, you could detach the inputs, allow metadata changes
on them, and then reflect metadata changes to the originals
as necessary (in fact, we should do this for other reasons),
but AOTAutograd doesn't do this yet.

Another alternative is to have dynamo guarantee not to give
duplicate tensor inputs, but because AOTAutograd is public API,
we are obligated to still handle it correctly here in case a
direct user passes duplicate inputs.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D39662821](https://our.internmc.facebook.com/intern/diff/D39662821)

ezyang committed:
Correctly handle duplicate arguments to AOTAutograd

If we do not deduplicate them, our custom autograd function
will double-count the gradient computed for the variable (since
the same x.grad field will be embedded into the graph twice.)

The alternative is to destroy aliasing relationships the inputs
and trace accumulating individual gradients for each of the
inputs into separate grad fields, but this prevents resizing
of inputs inside the traced graph from working correctly.  In
principle, you could detach the inputs, allow metadata changes
on them, and then reflect metadata changes to the originals
as necessary (in fact, we should do this for other reasons),
but AOTAutograd doesn't do this yet.

Another alternative is to have dynamo guarantee not to give
duplicate tensor inputs, but because AOTAutograd is public API,
we are obligated to still handle it correctly here in case a
direct user passes duplicate inputs.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Correctly handle duplicate arguments to AOTAutograd"


If we do not deduplicate them, our custom autograd function
will double-count the gradient computed for the variable (since
the same x.grad field will be embedded into the graph twice.)

The alternative is to destroy aliasing relationships the inputs
and trace accumulating individual gradients for each of the
inputs into separate grad fields, but this prevents resizing
of inputs inside the traced graph from working correctly.  In
principle, you could detach the inputs, allow metadata changes
on them, and then reflect metadata changes to the originals
as necessary (in fact, we should do this for other reasons),
but AOTAutograd doesn't do this yet.

Another alternative is to have dynamo guarantee not to give
duplicate tensor inputs, but because AOTAutograd is public API,
we are obligated to still handle it correctly here in case a
direct user passes duplicate inputs.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Correctly handle duplicate arguments to AOTAutograd"


If we do not deduplicate them, our custom autograd function
will double-count the gradient computed for the variable (since
the same x.grad field will be embedded into the graph twice.)

The alternative is to destroy aliasing relationships the inputs
and trace accumulating individual gradients for each of the
inputs into separate grad fields, but this prevents resizing
of inputs inside the traced graph from working correctly.  In
principle, you could detach the inputs, allow metadata changes
on them, and then reflect metadata changes to the originals
as necessary (in fact, we should do this for other reasons),
but AOTAutograd doesn't do this yet.

Another alternative is to have dynamo guarantee not to give
duplicate tensor inputs, but because AOTAutograd is public API,
we are obligated to still handle it correctly here in case a
direct user passes duplicate inputs.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

Chillee commented on functorch/_src/aot_autograd.py:
```diff
@@ -286,7 +286,58 @@ def new_fn(args):
 
 def aot_dispatch_autograd(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig):
-    joint_forward_backward = create_joint_forward_backward(flat_fn)
+    # Deduplicate inputs.  Suppose you have:
```
I don't think this should be done here. It should be done either in torchdynamo, or in `aot_function` (or `create_aot_dispatcher_function`) imo.

ezyang replied:
If you do it in `aot_function` you will miss the analogous bug with modules. create_aot_dispatcher_function is plausible, but you technically only need to do this in the autograd case so... meh?

See PR comment about torchdynamo.

eellison commented on functorch/_src/aot_autograd.py:
```diff
@@ -375,7 +426,7 @@ def backward(ctx, *flat_args):
             return tuple(out)
 
-    return CompiledFunction.apply
+    return lambda *args: CompiledFunction.apply(*remove_dupe_args(args))
```
nit: I don't use lambda here because it has worse stacktrace

eellison commented on functorch/_src/aot_autograd.py:
```diff
@@ -286,7 +286,58 @@ def new_fn(args):
+        deduped_flat_args.append(t)
+
+    def remove_dupe_args(args):
+        r = []
```
nit: since this is on the hot path, return `args` if `len(drop_args) == 0` ?

eellison commented on functorch/_src/aot_autograd.py:
```diff
@@ -286,7 +286,58 @@ def new_fn(args):
+    # else is just getting the types right.
+
+    seen_args = {}
+    deduped_flat_args = []
```
nit, I would maybe just call `deduped_flat_args = remove_dupe_args(args)` after this loop to make the loop do one less thing / consolidate logic

eellison commented on functorch/_src/aot_autograd.py:
```diff
@@ -286,7 +286,58 @@ def new_fn(args):
+    def remove_dupe_args(args):
+        r = []
+        for i, t in enumerate(args):
+            if i in drop_args:
```
nit, since this is on the hot path, can we do something like 

```
for t, keep_arg in zip(args, keep_args):
    if keep_arg:
         r.append(t)
```

eellison reviewed (commented):
Sweet, lgtm! Thank you for doing this. Will to defer to @Chillee or @anijain2305 for final accept

Chillee reviewed (approved):
Fine I'll approve, but I'd prefer if the logic was moved to `create_aot_autograd_function` :)

albanD reviewed (approved):
Sounds good

ezyang committed:
Update on "Correctly handle duplicate arguments to AOTAutograd"


If we do not deduplicate them, our custom autograd function
will double-count the gradient computed for the variable (since
the same x.grad field will be embedded into the graph twice.)

The alternative is to destroy aliasing relationships the inputs
and trace accumulating individual gradients for each of the
inputs into separate grad fields, but this prevents resizing
of inputs inside the traced graph from working correctly.  In
principle, you could detach the inputs, allow metadata changes
on them, and then reflect metadata changes to the originals
as necessary (in fact, we should do this for other reasons),
but AOTAutograd doesn't do this yet.

Another alternative is to have dynamo guarantee not to give
duplicate tensor inputs, but because AOTAutograd is public API,
we are obligated to still handle it correctly here in case a
direct user passes duplicate inputs.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Correctly handle duplicate arguments to AOTAutograd"


If we do not deduplicate them, our custom autograd function
will double-count the gradient computed for the variable (since
the same x.grad field will be embedded into the graph twice.)

The alternative is to destroy aliasing relationships the inputs
and trace accumulating individual gradients for each of the
inputs into separate grad fields, but this prevents resizing
of inputs inside the traced graph from working correctly.  In
principle, you could detach the inputs, allow metadata changes
on them, and then reflect metadata changes to the originals
as necessary (in fact, we should do this for other reasons),
but AOTAutograd doesn't do this yet.

Another alternative is to have dynamo guarantee not to give
duplicate tensor inputs, but because AOTAutograd is public API,
we are obligated to still handle it correctly here in case a
direct user passes duplicate inputs.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Correctly handle duplicate arguments to AOTAutograd"


If we do not deduplicate them, our custom autograd function
will double-count the gradient computed for the variable (since
the same x.grad field will be embedded into the graph twice.)

The alternative is to destroy aliasing relationships the inputs
and trace accumulating individual gradients for each of the
inputs into separate grad fields, but this prevents resizing
of inputs inside the traced graph from working correctly.  In
principle, you could detach the inputs, allow metadata changes
on them, and then reflect metadata changes to the originals
as necessary (in fact, we should do this for other reasons),
but AOTAutograd doesn't do this yet.

Another alternative is to have dynamo guarantee not to give
duplicate tensor inputs, but because AOTAutograd is public API,
we are obligated to still handle it correctly here in case a
direct user passes duplicate inputs.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39662821).
